lib/jxl/render_pipeline/simple_render_pipeline.cc

```diff
@@ -6,16 +6,146 @@
 #include "lib/jxl/render_pipeline/simple_render_pipeline.h"
 
 namespace jxl {
-void SimpleRenderPipeline::PrepareForThreads(size_t num) {
-  // TODO(veluca): actually allocate input buffers.
+void SimpleRenderPipeline::PrepareForThreadsInternal(size_t num) {
+  for (size_t c = 0; c < channel_shifts_[0].size(); c++) {
+    channel_data_.push_back(ImageF(DivCeil(frame_dimensions_.xsize_padded,
+                                           1 << channel_shifts_[0][c].first) +
+                                       kRenderPipelineXOffset * 2,
+                                   DivCeil(frame_dimensions_.ysize_padded,
+                                           1 << channel_shifts_[0][c].second) +
+                                       kRenderPipelineXOffset * 2));
+  }
 }
 
 std::vector<std::pair<ImageF*, Rect>> SimpleRenderPipeline::PrepareBuffers(
     size_t group_id, size_t thread_id) {
-  return {};
+  std::vector<std::pair<ImageF*, Rect>> ret;
+  for (size_t c = 0; c < channel_data_.size(); c++) {
+    const size_t gx = group_id % frame_dimensions_.xsize_groups;
+    const size_t gy = group_id / frame_dimensions_.xsize_groups;
+    const Rect rect(kRenderPipelineXOffset + gx * frame_dimensions_.group_dim,
+                    kRenderPipelineXOffset + gy * frame_dimensions_.group_dim,
+                    frame_dimensions_.group_dim, frame_dimensions_.group_dim,
+                    kRenderPipelineXOffset + frame_dimensions_.xsize_padded,
+                    kRenderPipelineXOffset + frame_dimensions_.ysize_padded);
+    ret.emplace_back(&channel_data_[c], rect);
+  }
+  return ret;
 }
 
 void SimpleRenderPipeline::ProcessBuffers(size_t group_id, size_t thread_id) {
-  // TODO(veluca): actually run the pipeline.
+  if (!IsDone()) return;
+  for (size_t stage_id = 0; stage_id < stages_.size(); stage_id++) {
+    const auto& stage = stages_[stage_id];
+    // Prepare buffers for kInOut channels.
+    std::vector<ImageF> new_channels(channel_data_.size());
+    std::vector<ImageF*> output_channels(channel_data_.size());
+
+    std::vector<std::pair<size_t, size_t>> input_sizes(channel_data_.size());
+    for (size_t c = 0; c < channel_data_.size(); c++) {
+      input_sizes[c] =
+          std::make_pair(channel_data_[c].xsize() - kRenderPipelineXOffset * 2,
+                         channel_data_[c].ysize() - kRenderPipelineXOffset * 2);
+    }
+
+    for (size_t c = 0; c < channel_data_.size(); c++) {
+      if (stage->GetChannelMode(c) != RenderPipelineChannelMode::kInOut) {
+        continue;
+      }
+      new_channels[c] =
+          ImageF((input_sizes[c].first << stage->settings_.shift_x) +
+                     kRenderPipelineXOffset * 2,
+                 (input_sizes[c].second << stage->settings_.shift_y) +
+                     kRenderPipelineXOffset * 2);
+      output_channels[c] = &new_channels[c];
+    }
+
+    // Add padding to all kInOut channels.
```

Comment:
This padding is to implement Mirroring like in section 6.5 of the spec, right? It might be useful to say that in the comment. In particular, I wouldn't really call this "padding", since to me that implies the values don't matter.